moped-editor/src/views/projects/projectView/ProjectComponents/agolUtils.js

```diff
@@ -51,6 +52,67 @@ const featureReducer = (geojson, { features, featureIdProp }) => {
   return { type: "FeatureCollection", features: uniqueFeatures };
 };
 
+/**
+ * Fetch CTN lines and points and a helper to find a feature record that matches a clicked layer feature by ID
+ * @param {String} linkMode - the current link mode ("lines" or "points")
+ * @param {Function} setIsFetchingFeatures - toggle loading state of the header spinner
+ * @param {Number} currentZoom - current level of zoom in the map
+ * @param {Array} bounds - the current map bounds
+ * @returns {AgolFeaturesObject}
+ */
+/**
+ * @typedef {Object} AgolFeaturesObject
+ * @property {Object} ctnLinesGeojson - Feature collection containing query results for CTN lines
+ * @property {Object} ctnPointsGeojson - Feature collection containing query results for CTN points
+ * @property {Function} findFeatureInAgolGeojsonFeatures - find a fetched AGOL feature record by unique ID
+ */
+export const useAgolFeatures = (
+  linkMode,
+  setIsFetchingFeatures,
+  currentZoom,
+  bounds
+) => {
+  const ctnLinesGeojson = useFeatureService({
+    layerId: SOURCES["ctn-lines"].featureService.layerId,
+    name: SOURCES["ctn-lines"].featureService.name,
+    bounds,
+    isVisible: linkMode === "lines" && currentZoom >= MIN_SELECT_FEATURE_ZOOM,
+    featureIdProp: SOURCES["ctn-lines"]._featureIdProp,
+    setIsFetchingFeatures,
+  });
+
+  const ctnPointsGeojson = useFeatureService({
+    layerId: SOURCES["ctn-points"].featureService.layerId,
+    name: SOURCES["ctn-points"].featureService.name,
+    bounds,
+    isVisible: linkMode === "points" && currentZoom >= MIN_SELECT_FEATURE_ZOOM,
+    featureIdProp: SOURCES["ctn-points"]._featureIdProp,
+    setIsFetchingFeatures,
+  });
+
+  const findFeatureInAgolGeojsonFeatures = (clickedFeature) => {
+    if (linkMode === "lines") {
+      const clickedFeatureId = clickedFeature.properties.CTN_SEGMENT_ID;
+
+      return ctnLinesGeojson.features.find(
+        (feature) => feature.properties.CTN_SEGMENT_ID === clickedFeatureId
+      );
+    } else if (linkMode === "points") {
+      const clickedFeatureId = clickedFeature.properties.INTERSECTION_ID;
+
+      return ctnPointsGeojson.features.find(
+        (feature) => feature.properties.INTERSECTION_ID === clickedFeatureId
+      );
+    }
+  };
```

Comment:
🙌 this looks good and works well. fwiw, i don't think we need to worry about our points layers being clipped, but if we for some reason moved our reference points into a to a multipoint geometry, it's possible clipping would occur. which is to say, i'm cool with playing it safe and handling clipping for points and lines.

@mddilley what do you think about nixing the hard-coded ID property - e.g. `clickedFeature.properties.CTN_SEGMENT_ID` in favor of looking that up from the configuration?

also, i kind of want to see this wrapped in `useCallback` or perhaps modified to accept the AGOL geojson layer as an argument and extracted completely from the hook? i don't feel strongly, and i liked seeing the feature service hooks combined into this wrapper, just not sure about cramming this new method.

Reply:
Yes! I like these ideas a lot. Definitely could get that config in here - I'll update that along with moving that helper outside of the hook.